scripts/update-message-index.py

```diff
@@ -0,0 +1,175 @@
+    description: str
+    details: str
+    added: Version
+    removed: Version | None
+    examples: tuple[Example, ...]
+
+
+def parse_version(raw_version: str) -> Version:
+    version = raw_version.split(".")
+    assert 2 <= len(version) <= 3 and all(n.isdecimal() for n in version), raw_version
+    return Version(*map(int, version))
+
+
+VALID_TYPES = ("warning", "error")
+DEFAULT_VERSION = "1.0"  # [TODO] some initial version?
+
+
+def parse_example(entry) -> Example:
+    name = entry.get("name")
+    assert name, entry
+    content = entry.get("content")
+    assert content
+    return Example(name=name, content=content)
+
+
+def parse_entry(entry) -> Entry:
+    code = entry.get("code")
+    assert code is not None, entry
+    id = entry.get("id")
+    assert id is not None, entry
+    type_ = entry.get("type")
+    assert type_ in VALID_TYPES, entry
+    description = entry.get("description")
+    assert description is not None, entry
+    details = entry.get("details", "")
+    added = parse_version(entry.get("added", DEFAULT_VERSION))
+    if removed := entry.get("removed"):
+        removed = parse_version(removed)
+    if examples := entry.get("examples", ()):
+        examples = tuple(map(parse_example, examples))
+    return Entry(
+        code=code,
+        id=id,
+        type=type_,
+        description=description,
+        added=added,
+        removed=removed,
+        details=details,
+        examples=examples,
+    )
+
+
+def skip_removed(entries: Sequence[Entry]):
+    return filter(lambda e: e.removed is None, entries)
+
+
+def make_message_code(entry: Entry) -> str:
+    return f"XKB-{entry.code:0>3}"
+
+
+def make_message_index(entry: Entry) -> str:
+    id = entry.id.replace("-", "_").upper()
+    return f"XKB_{entry.type.upper()}_{id}"
+
+
+def make_message_name(entry: Entry):
+    return entry.id.replace("-", " ").capitalize()
+
+
+def prepend_todo(text: str) -> str:
+    if text.startswith("TODO"):
+        return f"""<span class="TODO">{text[:5]}</span>{text[5:]}"""
+    else:
+        return text
+
+
+def generate_logging_h(index: Sequence[Entry], env: jinja2.Environment, root: Path):
+    template = env.get_template("src/messages.h.jinja")
+    path = root / "src" / "messages.h"
+    with path.open("wt", encoding="utf-8") as fd:
+        fd.writelines(template.generate(entries=skip_removed(index)))
+
+
+def generate_logging_c(index: Sequence[Entry], env: jinja2.Environment, root: Path):
+    template = env.get_template("src/messages.c.jinja")
+    path = root / "src" / "messages.c"
+    with path.open("wt", encoding="utf-8") as fd:
+        fd.writelines(template.generate(entries=skip_removed(index)))
+
+
+def generate_doc(index: Sequence[Entry], env: jinja2.Environment, root: Path):
+    path = root / "doc" / "message-index.md"
+    template = env.get_template("doc/message-index.md.jinja")
+    with path.open("wt", encoding="utf-8") as fd:
+        fd.writelines(template.generate(entries=index))
```

Comment:
nitpick: you could pass the filename into a more generic `generate()` function and either append `.jinja` or use `Path.stem()`. So basically:

```python
generate(message_index, jinja_env, args.output, root / "src" / "messages.h", skip_removed=True)
generate(message_index, jinja_env, args.output, root / "src" / "messages.c", skip_removed=True)
generate(message_index, jinja_env, args.output, root / "doc" / "message-index.md")
```